### PR TITLE
Introduce dynamic cluster metadata component

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Introduce dynamic cluster metadata component

**Why?**
To dynamically update cluster metadata

**How did you test it?**
Unit tests

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

File: common/cluster/dynamic_metadata.go [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+func (d *dynamicMetadataImpl) refreshClusterMetadata(_ context.Context) error {
+	clusterMetadataMap, err := d.listAllClusterMetadata()
+	if err != nil {
+		return err
+	}
+
+	oldEntries := make(map[string]*ClusterInformation)
+	newEntries := make(map[string]*ClusterInformation)
+
+	clusterInfoMap := d.GetAllClusterInfo()
```
Comment (yiminc) [inline common/cluster/dynamic_metadata.go:373]: it is confusing to see listAllClusterMetadata() and GetAllClusterInfo(), is list from db and Get from memory cache?

File: common/cluster/dynamic_metadata.go [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+		if !ok {
+			// handle new cluster registry
+			oldEntries[clusterName] = nil
+			newEntries[clusterName] = &ClusterInformation{
+				Enabled:                clusterMetadata.Enabled,
+				InitialFailoverVersion: clusterMetadata.InitialFailoverVersion,
+				RPCAddress:             clusterMetadata.RPCAddress,
+				version:                clusterMetadata.version,
+			}
+			continue UpdateLoop
```
Comment (yiminc) [inline common/cluster/dynamic_metadata.go:386]: we can remove this line and use else case at line 387

File: common/cluster/dynamic_metadata.go [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+		return err
+	}
+
+	oldEntries := make(map[string]*ClusterInformation)
+	newEntries := make(map[string]*ClusterInformation)
+
+	clusterInfoMap := d.GetAllClusterInfo()
+UpdateLoop:
+	for clusterName, clusterMetadata := range clusterMetadataMap {
+		clusterInfo, ok := clusterInfoMap[clusterName]
```
Comment (yiminc) [inline common/cluster/dynamic_metadata.go:376]: rename clusterMetadata -> newClusterInfo, and clusterInfo -> oldClusterInfo

File: common/cluster/dynamic_metadata.go [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+			}
+			newEntries[clusterName] = &ClusterInformation{
+				Enabled:                clusterMetadata.Enabled,
+				InitialFailoverVersion: clusterMetadata.InitialFailoverVersion,
+				RPCAddress:             clusterMetadata.RPCAddress,
+				version:                clusterMetadata.version,
+			}
+		}
+	}
+	for clusterName, clusterInfo := range clusterInfoMap {
```
Comment (yiminc) [inline common/cluster/dynamic_metadata.go:405]: clusterInfo -> oldClusterInfo

File: common/cluster/dynamic_metadata.go [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+			d.clusterInfo[clusterName] = *newClusterMetadata[clusterName]
+		}
+	}
+}
+
+func (d *dynamicMetadataImpl) updateVersionToClusterNameLocked() {
+	d.versionToClusterName = updateVersionToClusterName(d.clusterInfo, d.failoverVersionIncrement)
+}
+
+func updateVersionToClusterName(clusterInfo map[string]ClusterInformation, failoverVersionIncrement int64) map[int64]string {
```
Comment (yiminc) [inline common/cluster/dynamic_metadata.go:445]: updateVersionToClusterName -> updateFailoverVersionToClusterName

**Review by yiminc — APPROVED** [review]

**Comment by alexshtin** [comment]
2208 must be mine!

File: common/cluster/dynamic_metadata.go [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+func (d *dynamicMetadataImpl) ClusterNameForFailoverVersion(failoverVersion int64) string {
+	if failoverVersion == common.EmptyVersion {
+		return d.currentClusterName
+	}
+
+	initialFailoverVersion := failoverVersion % d.failoverVersionIncrement
+	// Failover version starts with 1.  Zero is an invalid value for failover version
+	if initialFailoverVersion == common.EmptyVersion {
+		initialFailoverVersion = d.failoverVersionIncrement
+	}
```
Comment (wxing1292) [inline common/cluster/dynamic_metadata.go:291]: why not simply panic?

File: common/cluster/dynamic_metadata.go [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+				case <-ctx.Done():
+					return nil
+				default:
+					d.logger.Error("Error refreshing remote cluster metadata", tag.Error(err))
+					select {
+					case <-time.After(refreshFailureInterval):
+					case <-ctx.Done():
+						return nil
+					}
+				}
```
Comment (wxing1292) [inline common/cluster/dynamic_metadata.go:358]:
```go
select {
case <-time.After(refreshFailureInterval):
case <-ctx.Done():
	return nil
}
```

File: common/cluster/dynamic_metadata.go (hunk 2 above)
Comment (yux0) [reply common/cluster/dynamic_metadata.go:386]: Good point. Updated

File: common/cluster/dynamic_metadata.go (hunk 1 above)
Comment (yux0) [reply common/cluster/dynamic_metadata.go:373]: Yes. I will update the private method name

File: common/cluster/dynamic_metadata.go (hunk 6 above)
Comment (yux0) [reply common/cluster/dynamic_metadata.go:291]: This is the original logic. I guess we allow 0 failover version in some previous release and to make it compatible?

**Pull request merged**